scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,617 @@
+import math
+import heapq
+import itertools
+import functools
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import numpy as np
+
+from numpy.polynomial.polynomial import Polynomial
+
+
+def cub(f, a, b, rule, rtol=1e-05, atol=1e-08, limit=10000, args=()):
+    if limit is None:
+        limit = np.inf
+
+    est = rule.estimate(f, a, b, args)
+    err = rule.error_estimate(f, a, b, args)
+
+    regions = [CubatureRegion(est, err, a, b)]
+    subdivisions = 0
+    success = False
+
+    while np.any(err > atol + rtol * np.abs(est)):
+        region = heapq.heappop(regions)
+
+        est_k = region.estimate
+        err_k = region.error
+
+        a_k, b_k = region.a, region.b
+        m_k = (a_k + b_k) / 2
+
+        subregion_coordinates = zip(
+            _cartesian_product(np.array([a_k, m_k]).T).T,
+            _cartesian_product(np.array([m_k, b_k]).T).T
+        )
+
+        est -= est_k
+        err -= err_k
+
+        for a_k_sub, b_k_sub in subregion_coordinates:
+            est_sub = rule.estimate(f, a_k_sub, b_k_sub, args)
+            err_sub = rule.error_estimate(f, a_k_sub, b_k_sub, args)
+
+            est += est_sub
+            err += err_sub
+
+            new_region = CubatureRegion(est_sub, err_sub, a_k_sub, b_k_sub)
+
+            heapq.heappush(regions, new_region)
+
+        subdivisions += 1
+
+        if subdivisions >= limit:
+            break
+    else:
+        success = True
+
+    status = "converged" if success else "not_converged"
+
+    return CubatureResult(
+        estimate=est,
+        error=err,
+        success=success,
+        status=status,
+        subdivisions=subdivisions,
+        regions=regions,
+        atol=atol,
+        rtol=rtol,
+    )
+
+
+@dataclass
+class CubatureRegion:
+    estimate: np.ndarray
+    error: np.ndarray
+    a: np.ndarray
+    b: np.ndarray
+
+    def __lt__(self, other):
+        # Consider regions with higher error as smaller so that they are placed at the
+        # top of the min-heap.
+        return _max_norm(self.error) > _max_norm(other.error)
+
+
+@dataclass
+class CubatureResult:
+    estimate: np.ndarray
+    error: np.ndarray
+    success: bool
+    status: str
+    regions: list[CubatureRegion]
+    subdivisions: int
+    atol: float
+    rtol: float
+
+
+class CubatureRule(ABC):
```

Comment:
Minor: Inheriting from ABC changes the MRO; I have vague recollections of this having some strange effects on isinstance / issubclass results down the line. Can't remember the details, so maybe it's FUD.